Source/WebGPU/WebGPU/Sampler.mm

```diff
@@ -272,41 +289,62 @@ static uint64_t floatToUint64(float f)
     return !!m_samplerIdentifier;
 }
 
+NSUInteger Sampler::aliveCount()
+{
+    Locker locker { samplerStateLock };
+    [weakSamplerStates _markNeedsCompaction];
+    [weakSamplerStates compact];
+    return weakSamplerStates.count;
+}
+
 id<MTLSamplerState> Sampler::samplerState() const
 {
     if (!m_samplerIdentifier)
         return nil;
 
-    Locker locker { samplerStateLock };
-    if (!cachedSamplerStates) {
-        cachedSamplerStates = [NSMutableDictionary dictionary];
-        lastAccessedKeys = [NSMutableOrderedSet orderedSet];
+    {
+        Locker locker { samplerStateLock };
+        if (!cachedSamplerStates) {
+            cachedSamplerStates = [NSMutableDictionary dictionary];
+            lastAccessedKeys = [NSMutableOrderedSet orderedSet];
+            weakSamplerStates = [NSPointerArray weakObjectsPointerArray];
+        }
+
+        id<MTLSamplerState> samplerState = [cachedSamplerStates objectForKey:m_samplerIdentifier];
+        if (samplerState) {
+            m_cachedSamplerState = samplerState;
+            return samplerState;
+        }
     }
 
-    id<MTLSamplerState> samplerState = [cachedSamplerStates objectForKey:m_samplerIdentifier];
-    if (samplerState)
-        return samplerState;
-
     id<MTLDevice> device = m_device->device();
     if (!device)
         return nil;
-    if (cachedSamplerStates.count >= device.maxArgumentBufferSamplerCount) {
+
+    if (aliveCount() >= device.maxArgumentBufferSamplerCount) {
+        Locker locker { samplerStateLock };
```

Comment:
The locking here is dubious. You do:
- lock
- free
- unlock
- allocate

If the locking is truely needed, then somebody can race free/alloc.

Consider algorithm of something like:

```
lock
do (
   alloc()
   if succeed break;
   free();
) while (can free)
unlock
```

If contention is an issue, try moving the lock inside.